### PR TITLE
[PM-8455] [PM-7683] Dynamic list items - Copy Action

### PR DESCRIPTION
**Pull request by shane-melton**
## 🎟️ Tracking

[PM-7683](https://bitwarden.atlassian.net/browse/PM-7683) and [PM-8455](https://bitwarden.atlassian.net/browse/PM-8455)

## 📔 Objective

Make the new browser vault list items more dynamic by adding the Copy Action menu that adapts to the cipher item type and available fields.

## 📚 `libs/vault` Additions

Add a `CopyCipherFieldService` that provides a `copy` method that will take care of copying a field value from a cipher. Internally it handles:
- Password re-prompting when applicable
- TOTP code generation if a available to the user/organization
- Showing an internationalized success toast
- Logging/collecting events when certain fields are copied to the users clipboard

Additionally, this introduces a new `CopyCipherFieldDirective` (`appCopyField`) to add functionality to a button to copy a cipher's field to the user's clipboard.

Example usage:
```html
<button type="button" bitMenuItem appCopyField="username" [cipher]="cipher">
  {{ "copyUsername" | i18n }}
</button>
```

## 📸 Screenshots

https://github.com/bitwarden/clients/assets/8764515/2593d185-0a2a-4afa-b4ed-f7c32047d492

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-7683]: https://bitwarden.atlassian.net/browse/PM-7683?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PM-8455]: https://bitwarden.atlassian.net/browse/PM-8455?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/vault/src/services/copy-cipher-field.service.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    if (
+      !skipReprompt &&
+      cipher.reprompt !== CipherRepromptType.None &&
+      action.protected &&
+      !(await this.passwordRepromptService.showPasswordPrompt())
+    ) {
+      return;
+    }
+
+    if (valueToCopy == null || !cipher.viewPassword) {
```
Comment (gbubemismith) [inline libs/vault/src/services/copy-cipher-field.service.ts:109]: I think we can combine this check here 

```suggestion
    if (valueToCopy == null || !cipher.viewPassword || (actionType === "totp" && !(await this.totpAllowed(cipher)))) {
```

**Review by gbubemismith — COMMENTED** [review]
This is looking really good. Thanks @shane-melton, I added a few suggestions.

File: libs/vault/src/services/copy-cipher-field.service.ts (hunk 1 above)
Comment (shane-melton) [reply libs/vault/src/services/copy-cipher-field.service.ts:109]: We could combine it, but we'd still need a separate check for `actionType === "totp"` in order to use the `totpService.getCode()` method below. I personally think its a little more clear to keep the `totp` "branch" separate  and have fewer conditions in a single `if` even if it repeats the empty `return` below.

**Review by gbubemismith — APPROVED** [review]
Nice job @shane-melton  🚀

**Pull request merged**